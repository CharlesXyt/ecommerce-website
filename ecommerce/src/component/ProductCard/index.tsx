import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import { Button, Chip } from '@mui/material';
import { Product } from '../../api/apiType';

export interface ProductCardProps extends Product {
	btnText: string;
	onClick: (id: number) => void;
}

export default function ProductCard(props: ProductCardProps) {
	const {
		id,
		imgSrc,
		imgAltText,
		heading,
		description,
		category,
		price,
		btnText,
		onClick,
	} = props;
	return (
		<Card
			sx={{
				maxWidth: 345,
				margin: 'auto',
				height: '100%',
				position: 'relative',
			}}
		>
			<CardMedia
				component='img'
				height='140'
				image={imgSrc}
				alt={imgAltText}
			/>
			<CardContent>
				<Typography gutterBottom variant='h5' component='div'>
					{heading}
				</Typography>
				<Typography variant='body2' color='text.secondary'>
					<span>{description}</span>
					<span>${price}</span>
				</Typography>
				<Button onClick={() => onClick(id)} color='inherit'>
					{btnText}
				</Button>
			</CardContent>
			<Chip
				sx={{
					position: 'absolute',
					top: '10px',
					right: '10px',
					padding: 1,
					backgroundColor: 'primary.main',
					color: 'primary.contrastText',
					borderRadius: '15px',
				}}
				label={category}
			/>
		</Card>
	);
}
