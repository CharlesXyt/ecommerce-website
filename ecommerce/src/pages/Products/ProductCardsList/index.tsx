import { Product } from '../../../api/apiType';
import ProductCard from '../../../component/ProductCard';
import { useCart } from '../../../context/CartContext';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import Notification from '../../../component/Notification';
import { useState } from 'react';

interface ProductCardsListProps {
	products: Product[];
}

const ProductCardsList = (props: ProductCardsListProps) => {
	const { products } = props;
	const { addToCart } = useCart();
	const [alert, setAlert] = useState(false);
	const addToCartHandler = (id: number) => {
		addToCart(id, 1);
		setAlert(true);
	};
	return (
		<Box sx={{ flexGrow: 1 }}>
			<Grid
				container
				spacing={{ xs: 2, md: 3 }}
				columns={{ xs: 4, sm: 8, md: 12 }}
			>
				{products.map((product, index) => (
					<Grid item xs={4} sm={4} md={3} key={index}>
						<ProductCard
							key={index}
							{...product}
							btnText='Add to Cart'
							onClick={addToCartHandler}
						/>
					</Grid>
				))}
			</Grid>
			<Notification
				onClose={() => {
					setAlert(false);
				}}
				showNotification={alert}
				message='Product added to cart successfully!'
				type='success'
			/>
		</Box>
	);
};

export default ProductCardsList;
