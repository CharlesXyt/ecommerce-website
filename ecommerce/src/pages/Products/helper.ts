type DebouncedFunction<F extends (...args: any[]) => any> = (
	...args: Parameters<F>
) => Promise<ReturnType<F>>;

export function debounce<F extends (...args: any[]) => any>(
	func: F,
	delay: number,
): DebouncedFunction<F> {
	let timeoutId: NodeJS.Timeout;

	return function (this: any, ...args: Parameters<F>) {
		return new Promise(resolve => {
			clearTimeout(timeoutId);

			timeoutId = setTimeout(() => {
				const result = func.apply(this, args);
				resolve(result);
			}, delay);
		});
	};
}
