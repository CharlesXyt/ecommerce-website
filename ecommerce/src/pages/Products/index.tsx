import { useEffect, useState, useCallback } from 'react';
import ProductCardsList from './ProductCardsList';
import Cart from '../../component/Cart';
import { useCart } from '../../context/CartContext';
import SearchInput from '../../component/SearchInput';
import _debounce from 'lodash/debounce';
import Filter from '../../component/Filter';
import { ProductCategory } from './models';

const Products = () => {
	const [products, setProducts] = useState([]);
	const [categories, setCategories] = useState([]);
	const [searchTerm, setSearchTerm] = useState('');
	const [filterTerm, setFilterTerm] = useState('');
	const { cart } = useCart();

	// unit test
	const productCount = cart.reduce((acc, item) => acc + item.quantity, 0);

	useEffect(() => {
		const fetchProducts = async () => {
			const response = await fetch('http://localhost:3000/products');
			const responseJson = await response.json();
			setProducts(responseJson.data);
		};
		const fetchCategories = async () => {
			const response = await fetch(
				'http://localhost:3000/products/categories',
			);
			const responseJson = await response.json();
			setCategories(responseJson);
		};
		try {
			Promise.all([fetchProducts(), fetchCategories()]);
		} catch (error) {
			console.error('Error fetching products', error);
		}
	}, []);

	const debounceSearch = useCallback(
		_debounce((searchTerm: string, filterTerm: string) => {
			const fetchData = async () => {
				console.log('filterTerm: ' + filterTerm);
				const response = await fetch(
					`http://localhost:3000/products?search=${searchTerm}&category=${filterTerm}`,
				);
				const responseJson = await response.json();
				setProducts(responseJson.data);
			};
			try {
				fetchData();
			} catch (error) {
				console.error('Error fetching products', error);
			}
		}, 300),
		[],
	);

	useEffect(() => {
		debounceSearch(searchTerm, filterTerm);
	}, [searchTerm, filterTerm, debounceSearch]);

	return (
		<div>
			<SearchInput
				handleSearch={event => setSearchTerm(event.target.value)}
			/>
			<Filter
				options={[
					{ label: 'Category 1', value: ProductCategory.ELECTRONICS },
					{ label: 'Category 2', value: ProductCategory.CLOTHING },
				]}
				handleFilter={value => setFilterTerm(value.join(','))}
			/>
			<Cart count={productCount} handleCartClick={() => {}} />
			<ProductCardsList products={products} />
		</div>
	);
};

export default Products;
