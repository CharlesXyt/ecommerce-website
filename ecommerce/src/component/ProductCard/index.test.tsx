import { render, screen, fireEvent } from '@testing-library/react';
import ProductCard from './index';
import { ProductCategory } from '../../pages/Products/models';

const mockProduct = {
	id: 1,
	imgSrc: 'image.jpg',
	imgAltText: 'Alt text',
	heading: 'Product 1',
	description: 'Description for Product 1',
	category: ProductCategory.ELECTRONICS,
	price: 29.99,
};

describe('ProductCard Component', () => {
	it('renders product card with correct information', () => {
		render(
			<ProductCard
				{...mockProduct}
				btnText='Add to Cart'
				onClick={() => {}}
			/>,
		);

		expect(screen.getByAltText('Alt text')).toBeInTheDocument();
		expect(screen.getByText('Product 1')).toBeInTheDocument();
		expect(
			screen.getByText('Description for Product 1'),
		).toBeInTheDocument();
		expect(screen.getByText('$29.99')).toBeInTheDocument();
		expect(screen.getByText('Add to Cart')).toBeInTheDocument();
		expect(
			screen.getByText(ProductCategory.ELECTRONICS),
		).toBeInTheDocument();
	});

	it('calls onClick handler when button is clicked', () => {
		const onClickMock = jest.fn();
		render(
			<ProductCard
				{...mockProduct}
				btnText='Add to Cart'
				onClick={onClickMock}
			/>,
		);

		fireEvent.click(screen.getByText('Add to Cart'));
		expect(onClickMock).toHaveBeenCalledWith(1);
	});
});
