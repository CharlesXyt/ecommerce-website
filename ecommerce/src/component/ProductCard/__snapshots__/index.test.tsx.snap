// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ri6ub7-MuiPaper-root-MuiCard-root"
  >
    <button
      class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
      tabindex="0"
      type="button"
    >
      <img
        alt="Smartphone Image"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="140"
        src="path/to/smartphone.jpg"
      />
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Smartphone
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          High-performance smartphone with advanced features.
          electronics
          499.99
        </p>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
      />
    </button>
  </div>
</div>
`;
